File: starfightersquadronsweb/src/main/js/CombatAction.js
define(
        [ "AttackDice", "DamageDealer", "DefenseDice", "Phase", "Pilot", "RangeRuler", "ShipDestroyedAction",
                "TargetLock", "UpgradeCard" ],
        function(AttackDice, DamageDealer, DefenseDice, Phase, Pilot, RangeRuler, ShipDestroyedAction, TargetLock,
                UpgradeCard)
        {
            "use strict";
            function CombatAction(environment, adjudicator, attacker, attackerPosition, weapon, defender,
                    defenderPosition, callback)
            {
                InputValidator.validateNotNull("environment", environment);
                InputValidator.validateNotNull("adjudicator", adjudicator);
                InputValidator.validateNotNull("attacker", attacker);
                InputValidator.validateNotNull("attackerPosition", attackerPosition);
                InputValidator.validateNotNull("weapon", weapon);
                InputValidator.validateNotNull("defender", defender);
                InputValidator.validateNotNull("defenderPosition", defenderPosition);

                var that = this;

                this.environment = function()
                {
                    return environment;
                };

                this.adjudicator = function()
                {
                    return adjudicator;
                };

                this.attacker = function()
                {
                    return attacker;
                };

                this.attackerPosition = function()
                {
                    return attackerPosition;
                };

                this.weapon = function()
                {
                    return weapon;
                };

                this.defender = function()
                {
                    return defender;
                };

                this.defenderPosition = function()
                {
                    return defenderPosition;
                };

                var executionCount = 0;

                this.executionCount = function()
                {
                    return executionCount;
                };

                this.doIt = function()
                {
                    executionCount++;
                    var upgrade = weapon.upgrade();

                    if (upgrade)
                    {
                        if (upgrade.spendFocus)
                        {
                            attacker.focus().decrease();
                        }

                        if (upgrade.spendTargetLock)
                        {
                            var targetLock = attacker.findTargetLockByDefender(defender);

                            if (targetLock)
                            {
                                attacker.removeAttackerTargetLock(targetLock);
                                defender.removeDefenderTargetLock(targetLock);
                            }
                        }

                        if (upgrade.discardThisCard)
                        {
                            attacker.discardUpgrade(weapon.upgradeKey());
                        }
                    }

                    var range = RangeRuler.getRange(attacker, attackerPosition, defender, defenderPosition);
                    attacker.combatState().range(range);

                    if (weapon.ranges().vizziniContains(range))
                    {
                        attacker.combatState().combatAction(this);

                        // Roll attack dice.
                        var attackDiceCount = attacker.computeAttackDiceCount(environment, weapon, defender, range);
                        var attackDice = new AttackDice(attackDiceCount);

                        if (weapon.upgradeKey() === UpgradeCard.ADVANCED_PROTON_TORPEDOES)
                        {
                            attackDice.changeOneToValue(AttackDice.Value.BLANK, AttackDice.Value.FOCUS);
                            attackDice.changeOneToValue(AttackDice.Value.BLANK, AttackDice.Value.FOCUS);
                            attackDice.changeOneToValue(AttackDice.Value.BLANK, AttackDice.Value.FOCUS);
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.CONCUSSION_MISSILES)
                        {
                            attackDice.changeOneToValue(AttackDice.Value.BLANK, AttackDice.Value.HIT);
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.HEAVY_LASER_CANNON)
                        {
                            attackDice.changeAllToValue(AttackDice.Value.CRITICAL_HIT, AttackDice.Value.HIT);
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.MANGLER_CANNON)
                        {
                            attackDice.changeOneToValue(AttackDice.Value.HIT, AttackDice.Value.CRITICAL_HIT);
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.PROTON_TORPEDOES)
                        {
                            attackDice.changeOneToValue(AttackDice.Value.FOCUS, AttackDice.Value.CRITICAL_HIT);
                        }

                        if (attacker.pilotKey() === Pilot.POE_DAMERON && attacker.focus().count() > 0)
                        {
                            attackDice.changeOneToValue(AttackDice.Value.FOCUS, AttackDice.Value.HIT);
                        }

                        if (attacker.isUpgradedWith(UpgradeCard.LONE_WOLF) && attackDice.blankCount() > 0)
                        {
                            var tokens1 = environment.getFriendlyTokensAtRange(attacker, RangeRuler.ONE);
                            var tokens2 = environment.getFriendlyTokensAtRange(attacker, RangeRuler.TWO);

                            if (tokens1.length === 0 && tokens2.length === 0)
                            {
                                attackDice.rerollBlank();
                            }
                        }

                        attacker.combatState().attackDice(attackDice);
                        environment.phase(Phase.COMBAT_ROLL_ATTACK_DICE);

                        // Modify attack dice.
                        var agent = attacker.agent();
                        agent.getModifyAttackDiceAction(environment, adjudicator, attacker, attackDice, defender,
                                finishModifyAttackDice);
                    }
                };

                function finishModifyAttackDice(attackAction)
                {
                    LOGGER.debug("CombatAction.finishModifyAttackDice() attackAction = " + attackAction);

                    if (attackAction)
                    {
                        attackAction.doIt();
                    }

                    environment.phase(Phase.COMBAT_MODIFY_ATTACK_DICE);
                    var attackDice = attacker.combatState().attackDice();

                    // Roll defense dice.
                    var range = attacker.combatState().range();
                    var defenderDiceCount = defender.computeDefenseDiceCount(weapon, range);
                    var defenseDice = new DefenseDice(defenderDiceCount);

                    if (defender.pilotKey() === Pilot.LUKE_SKYWALKER && defenseDice.focusCount() > 0)
                    {
                        defenseDice.changeOneToValue(DefenseDice.Value.FOCUS, DefenseDice.Value.EVADE);
                    }

                    if (defender.pilotKey() === Pilot.POE_DAMERON && defender.focus().count() > 0)
                    {
                        defenseDice.changeOneToValue(DefenseDice.Value.FOCUS, DefenseDice.Value.EVADE);
                    }

                    if (defender.isUpgradedWith(UpgradeCard.AUTOTHRUSTERS) && defenseDice.blankCount() > 0)
                    {
                        var isBeyondRange2 = (range === undefined || range === RangeRuler.THREE);
                        var isOutsideFiringArc = !weapon.isDefenderInFiringArc(attackerPosition, weapon
                                .primaryFiringArc(), defender, defenderPosition);

                        if (isBeyondRange2 || isOutsideFiringArc)
                        {
                            defenseDice.changeOneToValue(DefenseDice.Value.BLANK, DefenseDice.Value.EVADE);
                        }
                    }

                    if (defender.isUpgradedWith(UpgradeCard.LONE_WOLF) && defenseDice.blankCount() > 0)
                    {
                        var tokens1 = environment.getFriendlyTokensAtRange(defender, RangeRuler.ONE);
                        var tokens2 = environment.getFriendlyTokensAtRange(defender, RangeRuler.TWO);

                        if (tokens1.length === 0 && tokens2.length === 0)
                        {
                            defenseDice.rerollBlank();
                        }
                    }

                    if (defender.reinforce().count() > 0)
                    {
                        // Add one evade result.
                        defenseDice.spendEvadeToken();
                    }

                    attacker.combatState().defenseDice(defenseDice);
                    environment.phase(Phase.COMBAT_ROLL_DEFENSE_DICE);

                    // Modify defense dice.
                    var defenderAgent = defender.agent();
                    defenderAgent.getModifyDefenseDiceAction(environment, adjudicator, attacker, attackDice, defender,
                            defenseDice, finishModifyDefenseDice);
                }

                function finishModifyDefenseDice(defenseAction)
                {
                    LOGGER.debug("defenseAction = " + defenseAction);

                    if (defenseAction)
                    {
                        defenseAction.doIt();
                    }

                    environment.phase(Phase.COMBAT_MODIFY_DEFENSE_DICE);
                    var defenseDice = attacker.combatState().defenseDice();

                    // Compare results.
                    // Deal damage.
                    var attackDice = attacker.combatState().attackDice();
                    var damageDealer = new DamageDealer(environment, attackDice.hitCount(), attackDice
                            .criticalHitCount(), defender, defenseDice.evadeCount());

                    // Four possibilities:
                    // computer vs computer: call finishDealDamage() once
                    // computer vs human: call finishDealDamage() once
                    // human vs computer: call finishDealDamage() once
                    // human vs human: call finishDealDamage() once
                    var attackerAgent = attacker.agent();
                    var defenderAgent = defender.agent();
                    var attackerIsComputerAgent = attackerAgent.isComputerAgent();
                    var defenderIsComputerAgent = defenderAgent.isComputerAgent();
                    var callbackFunction = function()
                    {
                        finishDealDamage(damageDealer);
                    };

                    if (attackerIsComputerAgent && defenderIsComputerAgent)
                    {
                        // Both computer agents.
                        setTimeout(callbackFunction, 1000);
                    }
                    else if (attackerIsComputerAgent && !defenderIsComputerAgent)
                    {
                        defenderAgent.dealDamage(environment, adjudicator, attacker, attackDice, defender, defenseDice,
                                damageDealer, callbackFunction);
                    }
                    else if (!attackerIsComputerAgent && defenderIsComputerAgent)
                    {
                        attackerAgent.dealDamage(environment, adjudicator, attacker, attackDice, defender, defenseDice,
                                damageDealer, callbackFunction);
                    }
                    else if (!attackerIsComputerAgent && !defenderIsComputerAgent)
                    {
                        // Both human agents.
                        attackerAgent.dealDamage(environment, adjudicator, attacker, attackDice, defender, defenseDice,
                                damageDealer, callbackFunction);
                    }

                    environment.phase(Phase.COMBAT_DEAL_DAMAGE);
                }

                function finishDealDamage(damageDealer)
                {
                    LOGGER.trace("CombatAction.finishDealDamage() start");

                    var isDefenderHit = (damageDealer.hits() + damageDealer.criticalHits() > 0);
                    attacker.combatState().isDefenderHit(isDefenderHit);
                    LOGGER.debug("isDefenderHit ? " + isDefenderHit);

                    if (isDefenderHit)
                    {
                        if (weapon.upgradeKey() === UpgradeCard.ADVANCED_HOMING_MISSILES)
                        {
                            defender.addCriticalDamage(environment.drawDamage());
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.FLECHETTE_CANNON)
                        {
                            defender.addDamage(environment.drawDamage());
                            if (!defender.isStressed())
                            {
                                defender.receiveStress();
                            }
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.ION_CANNON ||
                                weapon.upgradeKey() === UpgradeCard.ION_CANNON_TURRET)
                        {
                            defender.addDamage(environment.drawDamage());
                            defender.ion().increase();
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.ION_PULSE_MISSILES)
                        {
                            defender.addDamage(environment.drawDamage());
                            defender.ion().increase();
                            defender.ion().increase();
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.TWIN_LASER_TURRET)
                        {
                            defender.addDamage(environment.drawDamage());
                        }
                        else
                        {
                            damageDealer.dealDamage();
                        }

                        if (attacker.pilotKey() === Pilot.WHISPER)
                        {
                            attacker.focus().increase();
                        }

                        if (weapon.upgradeKey() === UpgradeCard.ASSAULT_MISSILES)
                        {
                            environment.getTokensAtRange(defender, RangeRuler.ONE).forEach(function(token)
                            {
                                token.addDamage(environment.drawDamage());
                            });
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.FLECHETTE_TORPEDOES && defender.hullValue() <= 4)
                        {
                            defender.receiveStress();
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.ION_TORPEDOES)
                        {
                            environment.getTokensAtRange(defender, RangeRuler.ONE).forEach(function(token)
                            {
                                token.ion().increase();
                            });
                        }
                        else if (weapon.upgradeKey() === UpgradeCard.PLASMA_TORPEDOES)
                        {
                            defender.shield().decrease();
                        }

                        if (defender.isUpgradedWith(UpgradeCard.STEALTH_DEVICE))
                        {
                            defender.upgradeKeys().vizziniRemove(UpgradeCard.STEALTH_DEVICE);
                        }
                    }
                    else
                    {
                        // Defender not hit.
                        if (attacker.isUpgradedWith(UpgradeCard.BOSSK))
                        {
                            if (!attacker.isStressed())
                            {
                                attacker.receiveStress();
                            }

                            attacker.focus().increase();
                            var targetLock = new TargetLock(attacker, defender);
                            attacker.addAttackerTargetLock(targetLock);
                            defender.addDefenderTargetLock(targetLock);
                        }
                    }

                    if (attacker.isUpgradedWith(UpgradeCard.TACTICIAN) &&
                            attacker.combatState().range() === RangeRuler.TWO)
                    {
                        var firingArc = attacker.pilot().shipTeam.ship.primaryFiringArc;

                        if (weapon.isDefenderInFiringArc(attackerPosition, firingArc, defender, defenderPosition))
                        {
                            defender.receiveStress();
                        }
                    }

                    var myDefender, myDefenderPosition;

                    if (defender.parent !== undefined)
                    {
                        myDefender = defender.parent;
                        myDefenderPosition = environment.getPositionFor(myDefender);
                    }
                    else
                    {
                        myDefender = defender;
                        myDefenderPosition = defenderPosition;
                    }

                    if (myDefender.isDestroyed())
                    {
                        if (attacker.pilotKey() === Pilot.IG_88A)
                        {
                            attacker.recoverShield();
                        }

                        var shipDestroyedAction = new ShipDestroyedAction(environment, myDefender, myDefenderPosition);
                        shipDestroyedAction.doIt();
                        var delay = 1500;
                        setTimeout(function()
                        {
                            if (callback)
                            {
                                callback();
                            }
                        }, delay);
                    }
                    else
                    {
                        if ((weapon.upgradeKey() === UpgradeCard.CLUSTER_MISSILES || weapon.upgradeKey() === UpgradeCard.TWIN_LASER_TURRET) &&
                                executionCount < 2)
                        {
                            that.doIt();
                        }
                        else
                        {
                            if (callback)
                            {
                                callback();
                            }
                        }
                    }

                    LOGGER.trace("CombatAction.finishDealDamage() end");
                }
            }

            return CombatAction;
        });

File: starfightersquadrons/src/main/java/org/vizzini/starfightersquadrons/CombatAction.java
package org.vizzini.starfightersquadrons;

import org.apache.commons.lang3.builder.ToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.vizzini.core.InputValidator;
import org.vizzini.core.game.Action;
import org.vizzini.starfightersquadrons.DamageCard.Trait;
import org.vizzini.starfightersquadrons.RangeRuler.Range;

/**
 * Provides a combat action for Starfighter Squadrons.
 */
public final class CombatAction implements Action
{
    /**
     * Provides a damage dealer for Starfighter Squadrons.
     */
    public static final class DamageDealer
    {
        /** Attacker critical hit count. */
        private final int criticalHitCount;

        /** Defender token. */
        private final SSToken defender;

        /** Environment. */
        private final SSEnvironment environment;

        /** Defender evade count. */
        private final int evadeCount;

        /** Attacker hit count. */
        private final int hitCount;

        /**
         * Construct this object.
         *
         * @param environment Environment.
         * @param hitCount Attacker hit count.
         * @param criticalHitCount Attacker critical hit count.
         * @param defender Defender token.
         * @param evadeCount Defender evade count.
         */
        @SuppressWarnings("hiding")
        public DamageDealer(final SSEnvironment environment, final int hitCount, final int criticalHitCount,
                final SSToken defender, final int evadeCount)
        {
            InputValidator.validateNotNull("environment", environment);

            this.environment = environment;
            this.hitCount = hitCount;
            this.criticalHitCount = criticalHitCount;
            this.defender = defender;
            this.evadeCount = evadeCount;
        }

        /**
         * Deal damage.
         */
        @SuppressWarnings("synthetic-access")
        public void dealDamage()
        {
            int hits = hitCount;
            int criticalHits = criticalHitCount;

            LOGGER.debug("hits         = " + hits);
            LOGGER.debug("criticalHits = " + criticalHits);

            int evades = evadeCount;

            LOGGER.debug("evades       = " + evades);

            if (hits > 0)
            {
                final int count = Math.min(hits, evades);
                hits -= count;
                evades -= count;
            }

            if (criticalHits > 0)
            {
                final int count = Math.min(criticalHits, evades);
                criticalHits -= count;
                evades -= count;
            }

            LOGGER.debug("final hits         = " + hits);
            LOGGER.debug("final criticalHits = " + criticalHits);
            LOGGER.debug("final evades       = " + evades);
            LOGGER.debug("before hits, shield                  = " + defender.getShieldCount());

            if (defender.getShieldCount() > 0)
            {
                final int count = Math.min(hits, defender.getShieldCount());
                hits -= count;

                for (int i = 0; i < count; i++)
                {
                    defender.decreaseShieldCount();
                }
            }

            LOGGER.debug("before critical hits, shield         = " + defender.getShieldCount());

            if (defender.getShieldCount() > 0)
            {
                final int count = Math.min(criticalHits, defender.getShieldCount());
                criticalHits -= count;

                for (int i = 0; i < count; i++)
                {
                    defender.decreaseShieldCount();
                }
            }

            LOGGER.debug("after both hits, shield              = " + defender.getShieldCount());
            LOGGER.debug("before hits, damage                  = " + defender.getDamageCount());

            for (int i = 0; i < hits; i++)
            {
                defender.addDamage(environment.drawDamage());
            }

            LOGGER.debug("after hits, damage                   = " + defender.getDamageCount());
            LOGGER.debug("before critical hits, criticalDamage = " + defender.getCriticalDamageCount());

            for (int i = 0; i < criticalHits; i++)
            {
                final DamageCard damage = environment.drawDamage();

                if (defender.isUpgradedWith(UpgradeCard.DETERMINATION) && (damage.getTrait() == Trait.PILOT))
                {
                    environment.discardDamage(damage);
                }
                else if (defender.getPilot() == Pilot.CHEWBACCA)
                {
                    defender.addDamage(damage);
                }
                else
                {
                    damage.dealEffect(environment, defender);
                }
            }

            LOGGER.debug("after critical hits, criticalDamage  = " + defender.getCriticalDamageCount());
            LOGGER.debug("defender.isDestroyed() ? " + defender.isDestroyed());
        }
    }

    /** Logger. */
    private static final Logger LOGGER = LogManager.getLogger();

    /** Adjudicator. */
    private final SSAdjudicator adjudicator;

    /** Attacker. */
    private final SSToken attacker;

    /** Defender. */
    private final SSToken defender;

    /** Environment. */
    private final SSEnvironment environment;

    /** Range ruler. */
    private final RangeRuler rangeRuler = new RangeRuler();

    /** Weapon. */
    private final Weapon weapon;

    /**
     * Construct this object.
     *
     * @param environment Environment.
     * @param adjudicator Adjudicator.
     * @param attacker Attacking token.
     * @param weapon Weapon.
     * @param defender Defending token.
     */
    @SuppressWarnings("hiding")
    public CombatAction(final SSEnvironment environment, final SSAdjudicator adjudicator,
            final SSToken attacker, final Weapon weapon, final SSToken defender)
    {
        InputValidator.validateNotNull("environment", environment);
        InputValidator.validateNotNull("adjudicator", adjudicator);
        InputValidator.validateNotNull("attacker", attacker);
        InputValidator.validateNotNull("weapon", weapon);
        InputValidator.validateNotNull("defender", defender);

        this.environment = environment;
        this.adjudicator = adjudicator;
        this.attacker = attacker;
        this.weapon = weapon;
        this.defender = defender;
    }

    @Override
    public boolean doIt()
    {
        attacker.setWeapon(weapon);

        final SSPosition attackerPosition = environment.getPositionFor(attacker);
        final SSPosition defenderPosition = environment.getPositionFor(defender);
        final Range range = rangeRuler.getRange(attacker, attackerPosition, defender, defenderPosition);
        LOGGER.trace("range = " + range);
        attacker.setRange(range);

        if (range != null)
        {
            LOGGER.trace("attacker = " + attacker);
            LOGGER.trace("defender = " + defender);
            attacker.setCombatAction(this);

            // Roll attack dice.
            final int attackDiceCount = attacker.computeAttackDiceCount(environment, weapon, range);
            final AttackDice attackDice = new AttackDice(attackDiceCount);
            attacker.setAttackDice(attackDice);
            environment.setPhase(Phase.COMBAT_ROLL_ATTACK_DICE);

            // Modify attack dice.
            final SSAgent agent = getAgent();
            ModifyAttackDiceAction attackAction = agent.getModifyAttackDiceAction(environment, adjudicator, attacker,
                    attackDice, defender);

            while (attackAction != null)
            {
                attackAction.doIt();
                attackAction = agent
                        .getModifyAttackDiceAction(environment, adjudicator, attacker, attackDice, defender);
            }

            environment.setPhase(Phase.COMBAT_MODIFY_ATTACK_DICE);
            LOGGER.trace("attackDice  = " + attackDice);

            // Roll defense dice.
            final int defenderDiceCount = defender.computeDefenseDiceCount(weapon, range);
            final DefenseDice defenseDice = new DefenseDice(defenderDiceCount);
            attacker.setDefenseDice(defenseDice);
            environment.setPhase(Phase.COMBAT_ROLL_DEFENSE_DICE);

            // Modify defense dice.
            final SSAgent defenderAgent = defender.getAgent();
            ModifyDefenseDiceAction defenseAction = defenderAgent.getModifyDefenseDiceAction(environment, adjudicator,
                    attacker, attackDice, defender, defenseDice);

            while (defenseAction != null)
            {
                defenseAction.doIt();
                defenseAction = defenderAgent.getModifyDefenseDiceAction(environment, adjudicator, attacker,
                        attackDice, defender, defenseDice);
            }

            environment.setPhase(Phase.COMBAT_MODIFY_DEFENSE_DICE);
            LOGGER.trace("defenseDice = " + defenseDice);

            // Compare results.
            // Deal damage.
            final int beforeDamage = defender.getDamageCount() + defender.getCriticalDamageCount();
            LOGGER.trace("beforeDamage = " + beforeDamage);
            final DamageDealer damageDealer = new DamageDealer(environment, attackDice.getHitCount(),
                    attackDice.getCriticalHitCount(), defender, defenseDice.getEvadeCount());
            damageDealer.dealDamage();
            final int afterDamage = defender.getDamageCount() + defender.getCriticalDamageCount();
            LOGGER.trace("afterDamage = " + afterDamage);
            final boolean isDefenderHit = afterDamage > beforeDamage;
            attacker.setDefenderHit(isDefenderHit);
            environment.setPhase(Phase.COMBAT_DEAL_DAMAGE);

            if (defender.isDestroyed())
            {
                final ShipDestroyedAction shipDestroyedAction = new ShipDestroyedAction(environment, defender,
                        defenderPosition);
                shipDestroyedAction.doIt();
            }
        }

        return false;
    }

    /**
     * @return the adjudicator
     */
    public SSAdjudicator getAdjudicator()
    {
        return adjudicator;
    }

    @Override
    public SSAgent getAgent()
    {
        return attacker.getAgent();
    }

    /**
     * @return the attacker
     */
    public SSToken getAttacker()
    {
        return attacker;
    }

    /**
     * @return the defender
     */
    public SSToken getDefender()
    {
        return defender;
    }

    @Override
    public SSEnvironment getEnvironment()
    {
        return environment;
    }

    /**
     * @return the weapon
     */
    public Weapon getWeapon()
    {
        return weapon;
    }

    @Override
    public String toString()
    {
        final ToStringBuilder builder = new ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE);

        builder.append("attacker", attacker);
        builder.append("defender", defender);

        return builder.toString();
    }

    @Override
    public boolean undoIt()
    {
        throw new RuntimeException("method not used");
    }
}
